Fenced code block causes segfault on Windows (`x64-mingw-ucrt`) since 2.8.0

Reproduction code:

`segfault-repro.rb`

```rb
require 'commonmarker'

md_text = "```\nHello, world\n```\n"
pp Commonmarker.to_html(md_text, plugins: { syntax_highlighter: { theme: "" } })
```

> [!NOTE]
> The `syntax_highlighter` plugin must be enabled for the segfault to occur, but it's enabled by default, so it's not necessary to configure it explicitly as above (that just makes the Rust extension code to take the simplest possible code path). Printing the result is not necessary either (that's only meaningful for the working versions to prove that there was no segfault), because the segfault happens inside `Commonmarker.to_html` before it returns. Therefore, `Commonmarker.to_html(md_text)` would also be sufficient to reproduce the segfault.

1. `2.7.0` as baseline (no segfault, works normally)

   `Gemfile`

   ```rb
   source 'https://rubygems.org'

   gem 'commonmarker', '2.7.0'
   ```

   <details>
   <summary><code>Gemfile.lock</code></summary>

   ```Gemfile.lock
   GEM
     remote: https://rubygems.org/
     specs:
       commonmarker (2.7.0-x64-mingw-ucrt)

   PLATFORMS
     x64-mingw-ucrt

   DEPENDENCIES
     commonmarker (= 2.7.0)

   CHECKSUMS
     commonmarker (2.7.0-x64-mingw-ucrt) sha256=e7d88dae657598303c0197d3fcb840a0f11283c043918c116348894b104d35b1

   BUNDLED WITH
     4.0.15
   ```
   </details>

   ```console
   $ ruby --version
   ruby 4.0.5 (2026-05-20 revision 64336ffd0e) +PRISM [x64-mingw-ucrt]

   $ gem update --system
   (...)

   $ gem --version
   4.0.15

   $ bundle --version
   4.0.15

   $ bundle install
   Fetching gem metadata from https://rubygems.org/.....
   Fetching commonmarker 2.7.0 (x64-mingw-ucrt)
   Installing commonmarker 2.7.0 (x64-mingw-ucrt)
   Bundle complete! 1 Gemfile dependency, 1 gem now installed.
   Use `bundle info [gemname]` to see where a bundled gem is installed.
   1 installed gem you directly depend on is looking for funding.
     Run `bundle fund` for details

   $ bundle exec ruby segfault-repro.rb
   "<pre class=\"syntax-highlighting\"><code><span class=\"text plain\">Hello, world\n" + "</span></code></pre>\n"
   ```

2. `2.8.0` - the first version where the segfault occurs

   `Gemfile`

   ```rb
   source 'https://rubygems.org'

   gem 'commonmarker', '2.8.0'
   ```

   <details>
   <summary><code>Gemfile.lock</code></summary>

   ```Gemfile.lock
   GEM
     remote: https://rubygems.org/
     specs:
       commonmarker (2.8.0-x64-mingw-ucrt)

   PLATFORMS
     x64-mingw-ucrt

   DEPENDENCIES
     commonmarker (= 2.8.0)

   CHECKSUMS
     commonmarker (2.8.0-x64-mingw-ucrt) sha256=5b68c1ba445a2c84abc98ae0f2e659c30088fc97b3817c16bb2eb345ec84c5d2

   BUNDLED WITH
     4.0.15
   ```
   </details>

   ```console
   WDAGUtilityAccount@8e95d2f3-261f-417a-8a06-5fbf6c031e74 MINGW64 /c/work/commonmarker-2.8.0-segfault-on-windows/cmarker-2.8.0-bad
   $ bundle install
   Fetching gem metadata from https://rubygems.org/.....
   Fetching commonmarker 2.8.0 (x64-mingw-ucrt)
   Installing commonmarker 2.8.0 (x64-mingw-ucrt)
   Bundle complete! 1 Gemfile dependency, 1 gem now installed.
   Use `bundle info [gemname]` to see where a bundled gem is installed.
   1 installed gem you directly depend on is looking for funding.
     Run `bundle fund` for details

   WDAGUtilityAccount@8e95d2f3-261f-417a-8a06-5fbf6c031e74 MINGW64 /c/work/commonmarker-2.8.0-segfault-on-windows/cmarker-2.8.0-bad
   $ bundle exec ruby segfault-repro.rb
   Segmentation fault         bundle exec ruby segfault-repro.rb

   WDAGUtilityAccount@8e95d2f3-261f-417a-8a06-5fbf6c031e74 MINGW64 /c/work/commonmarker-2.8.0-segfault-on-windows/cmarker-2.8.0-bad
   $ echo $?
   139
   ```

   Note that `Segmentation fault` is not really a Windows message - this is just how [Git for Windows](https://gitforwindows.org/) reports what would otherwise be a silent crash. A typical Windows behavior is just not to print anything in case of a crash like this and just set the exit code appropriately. In PowerShell 7.4+, [`$PSNativeCommandUseErrorActionPreference`](https://learn.microsoft.com/en-us/powershell/module/microsoft.powershell.core/about/about_error_handling?view=powershell-7.6#external-program-errors) can be set to report a non-zero code automatically as an error (in older PowerShell versions, `echo $LastExitCode` must be run explicitly after the failed command, `echo %errorlevel%` is an equivalent for `cmd.exe`):

   ```console
   PowerShell 7.6.3
   PS C:\Users\WDAGUtilityAccount> cd C:\work\commonmarker-2.8.0-segfault-on-windows\cmarker-2.8.0-bad
   PS C:\work\commonmarker-2.8.0-segfault-on-windows\cmarker-2.8.0-bad> $PSNativeCommandUseErrorActionPreference = $true
   PS C:\work\commonmarker-2.8.0-segfault-on-windows\cmarker-2.8.0-bad> bundle exec ruby segfault-repro.rb
   NativeCommandExitException: Program "bundle.bat" ended with non-zero exit code: -1073741819 (0xC0000005).
   ```

   The exit code `-1073741819 (0xC0000005)` means "Access Violation", i.e. a segfault.

   Sometimes (something like 1 in 20 run attempts?) Ruby catches the segfault and provides a best-effort `[BUG] Segmentation fault` report with backtraces, but it's not very helpful, because the `commonmarker.so` binary is stripped and has no debug symbols:

   <details>
   <summary><code>[BUG] Segmentation fault</code> backtraces by Ruby (not very helpful)</summary>

   ```console
   PS C:\work\commonmarker-2.8.0-segfault-on-windows\cmarker-2.8.0-bad> bundle exec ruby segfault-repro.rb
   C:/Ruby40-x64/lib/ruby/gems/4.0.0/gems/commonmarker-2.8.0-x64-mingw-ucrt/lib/commonmarker.rb:44: [BUG] Segmentation fault
   ruby 4.0.5 (2026-05-20 revision 64336ffd0e) +PRISM [x64-mingw-ucrt]

   -- Control frame information -----------------------------------------------
   c:0004 p:---- s:0023 e:000022 l:y b:---- CFUNC  :commonmark_to_html
   c:0003 p:0153 s:0017 e:000016 l:y b:0001 METHOD C:/Ruby40-x64/lib/ruby/gems/4.0.0/gems/commonmarker-2.8.0-x64-mingw-ucrt/lib/commonmarker.rb:44
   c:0002 p:0027 s:0008 E:0014f8 l:n b:---- EVAL   segfault-repro.rb:4 [FINISH]
   c:0001 p:0000 s:0003 E:000c60 l:y b:---- DUMMY  [FINISH]

   -- Ruby level backtrace information ----------------------------------------
   segfault-repro.rb:4:in '<main>'
   C:/Ruby40-x64/lib/ruby/gems/4.0.0/gems/commonmarker-2.8.0-x64-mingw-ucrt/lib/commonmarker.rb:44:in 'to_html'
   C:/Ruby40-x64/lib/ruby/gems/4.0.0/gems/commonmarker-2.8.0-x64-mingw-ucrt/lib/commonmarker.rb:44:in 'commonmark_to_html'

   -- Threading information ---------------------------------------------------
   Total ractor count: 1
   Ruby thread count for this ractor: 1

   -- C level backtrace information -------------------------------------------
   C:\Windows\SYSTEM32\ntdll.dll(ZwWaitForSingleObject+0x14) [0x00007fff2848d6d4]
   C:\Windows\System32\KERNELBASE.dll(WaitForSingleObjectEx+0x8e) [0x00007fff25a7b6ae]
   C:\Ruby40-x64\bin\x64-ucrt-ruby400.dll(rb_vm_bugreport+0x237) [0x00007fff05a558e7]
   C:\Ruby40-x64\bin\x64-ucrt-ruby400.dll(rb_bug_for_fatal_signal+0x9b) [0x00007fff057f032b]
   C:\Ruby40-x64\bin\x64-ucrt-ruby400.dll(rb_shape_free_all+0xfb5) [0x00007fff059856c5]
    [0x00007ff72a9c20a2]
   C:\Windows\System32\ucrtbase.dll(_C_specific_handler+0xa0) [0x00007fff26290e10]
   C:\Windows\SYSTEM32\ntdll.dll(_chkstk+0x11f) [0x00007fff28492a1f]
   C:\Windows\SYSTEM32\ntdll.dll(RtlRaiseException+0x484) [0x00007fff28442554]
   C:\Windows\SYSTEM32\ntdll.dll(KiUserExceptionDispatcher+0x2e) [0x00007fff2849152e]
   C:\Ruby40-x64\lib\ruby\gems\4.0.0\gems\commonmarker-2.8.0-x64-mingw-ucrt\lib\commonmarker\4.0\commonmarker.so(ruby_abi_version+0x6544f) [0x0000000074cac07f]
   C:\Ruby40-x64\lib\ruby\gems\4.0.0\gems\commonmarker-2.8.0-x64-mingw-ucrt\lib\commonmarker\4.0\commonmarker.so(ruby_abi_version+0x80d03) [0x0000000074cc7933]
   C:\Ruby40-x64\lib\ruby\gems\4.0.0\gems\commonmarker-2.8.0-x64-mingw-ucrt\lib\commonmarker\4.0\commonmarker.so(ruby_abi_version+0x72ff0) [0x0000000074cb9c20]
   C:\Ruby40-x64\lib\ruby\gems\4.0.0\gems\commonmarker-2.8.0-x64-mingw-ucrt\lib\commonmarker\4.0\commonmarker.so(ruby_abi_version+0x47208) [0x0000000074c8de38]
   C:\Ruby40-x64\lib\ruby\gems\4.0.0\gems\commonmarker-2.8.0-x64-mingw-ucrt\lib\commonmarker\4.0\commonmarker.so(ruby_abi_version+0x44ca6) [0x0000000074c8b8d6]
   C:\Ruby40-x64\lib\ruby\gems\4.0.0\gems\commonmarker-2.8.0-x64-mingw-ucrt\lib\commonmarker\4.0\commonmarker.so(Init_commonmarker+0x3244d) [0x0000000074be381d]
   C:\Ruby40-x64\lib\ruby\gems\4.0.0\gems\commonmarker-2.8.0-x64-mingw-ucrt\lib\commonmarker\4.0\commonmarker.so(Init_commonmarker+0x3c1be) [0x0000000074bed58e]
   C:\Ruby40-x64\lib\ruby\gems\4.0.0\gems\commonmarker-2.8.0-x64-mingw-ucrt\lib\commonmarker\4.0\commonmarker.so(Init_commonmarker+0x7559e) [0x0000000074c2696e]
   C:\Ruby40-x64\lib\ruby\gems\4.0.0\gems\commonmarker-2.8.0-x64-mingw-ucrt\lib\commonmarker\4.0\commonmarker.so(Init_commonmarker+0x6626) [0x0000000074bb79f6]
   C:\Ruby40-x64\bin\x64-ucrt-ruby400.dll(rb_error_arity+0x130) [0x00007fff05a1e5b0]
   C:\Ruby40-x64\bin\x64-ucrt-ruby400.dll(rb_error_arity+0x42c) [0x00007fff05a1e8ac]
   C:\Ruby40-x64\bin\x64-ucrt-ruby400.dll(rb_ec_obj_respond_to+0x932) [0x00007fff05a43db2]
   C:\Ruby40-x64\bin\x64-ucrt-ruby400.dll(rb_vm_exec+0xf4) [0x00007fff05a32904]
   C:\Ruby40-x64\bin\x64-ucrt-ruby400.dll(rb_refinement_module_get_refined_class+0x2e56) [0x00007fff057fb976]
   C:\Ruby40-x64\bin\x64-ucrt-ruby400.dll(ruby_run_node+0x70) [0x00007fff057fdfd0]
    [0x00007ff72a9c1503]
    [0x00007ff72a9c10da]
    [0x00007ff72a9c1456]
   C:\Windows\System32\KERNEL32.DLL(BaseThreadInitThunk+0x14) [0x00007fff26e47374]

   -- Other runtime information -----------------------------------------------

   * Loaded script: segfault-repro.rb

   * Ruby Box: disabled
   * Loaded features:

     (...)

   NativeCommandExitException: Program "bundle.bat" ended with non-zero exit code: 3 (0x00000003).
   ```
   </details>

---

I wanted to find out more details so I could submit as helpful a bug report as possible, so I dug deeper.

First, I reviewed the https://github.com/gjtorikian/commonmarker/compare/v2.7.0...v2.8.0 diff ([permalink](https://github.com/gjtorikian/commonmarker/compare/863a6790b51e77ca8ff15d3f4082be8ae87f9df5...3ad639093fee4d36798afc1dac615d423c60a792)), but nothing immediately struck me as suspicious (in hindsight, I can say that the clues were there, but more on that later).

Of course, I tried the latest version 2.8.2 (https://rubygems.org/gems/commonmarker/versions/2.8.2-x64-mingw-ucrt) in case the bug had been fixed, but nothing changed.

I decided to build `commonmarker` locally from source with debug information. FWIW, I did this on Windows 10 in a Windows Sandbox session - for my own reference, here are the steps I followed:

<details>
<summary>Steps to build <code>commonmarker</code> locally on Windows</summary>

#### Setup

1. Download "x86_64-pc-windows-gnu" from https://rust-lang.github.io/rustup/installation/other.html
2. Verify integrity

   ```console
   PS C:\Users\WDAGUtilityAccount\Downloads> certutil -hashfile .\rustup-init.exe SHA256
   SHA256 hash of .\rustup-init.exe:
   03dbaae1d33a4d220bd0d202e5092955dae859c119074192ca513f8c4713fff7
   CertUtil: -hashfile command completed successfully.
   PS C:\Users\WDAGUtilityAccount\Downloads> cat .\rustup-init.exe.sha256
   03dbaae1d33a4d220bd0d202e5092955dae859c119074192ca513f8c4713fff7
   ```

3. Install the GNU toolchain (this is how the `commonmarker` gem for the `x64-mingw-ucrt` platform is normally built)

   ```console
   PS C:\Users\WDAGUtilityAccount\Downloads> .\rustup-init.exe -y --default-host x86_64-pc-windows-gnu --default-toolchain stable --profile minimal
   ```

4. Install Ruby 4.0.5 with Devkit

   ```powershell
   winget install -i -e --id RubyInstallerTeam.RubyWithDevKit.4.0
   ```

   Installation folder: C:\Ruby40-x64

   - [x] Add Ruby executables to your PATH
   - [ ] Associate .rb and .rbw files with this Ruby installation

   ---

   Select Components

   - [x] Ruby-4.0.5 base files (56.5 MB)
   - [ ] Ruby RI and HTML documentation (214.6 MB)
   - [x] MSYS2 development toolchain 2026-05-20 (899.8 MB)

   ---

   - [x] Run 'ridk install' to set up MSYS2 and development toolchain.

   ---

   ```
      1 - MSYS2 base installation
      2 - MSYS2 system update (optional)
      3 - MSYS2 and MINGW development toolchain

   Which components shall be installed? If unsure press ENTER [1,3] <press Enter>
   ```

5. Run `gem update --system` to update Bundler

6. Run these commands in Command Prompt (`cmd.exe`):

   ```cmd
   ridk enable
   pacman -S mingw-w64-ucrt-x86_64-clang
   ```

7. Clone this repo (https://github.com/gjtorikian/commonmarker), run `bundle install`

   ```console
   WDAGUtilityAccount@8e95d2f3-261f-417a-8a06-5fbf6c031e74 MINGW64 /c/work/commonmarker (windows-segfault-repro)
   $ bundle check
   The following gems are missing
    * rake (13.4.2)
    * rake-compiler (1.3.1)
    * rb_sys (0.9.128)
    * rake-compiler-dock (1.12.0)
   Install missing gems with `bundle install`

   WDAGUtilityAccount@8e95d2f3-261f-417a-8a06-5fbf6c031e74 MINGW64 /c/work/commonmarker (windows-segfault-repro)
   $ bundle install
   Fetching gem metadata from https://rubygems.org/........
   Fetching rake 13.4.2
   Installing rake 13.4.2
   Fetching rake-compiler-dock 1.12.0
   Fetching rb_sys 0.9.128
   Fetching rake-compiler 1.3.1
   Installing rake-compiler 1.3.1
   Installing rb_sys 0.9.128
   Installing rake-compiler-dock 1.12.0
   Bundle complete! 13 Gemfile dependencies, 5 gems now installed.
   Gems in the groups 'debug', 'test', 'lint' and 'benchmark' were not installed.
   Use `bundle info [gemname]` to see where a bundled gem is installed.
   1 installed gem you directly depend on is looking for funding.
     Run `bundle fund` for details
   ```

#### Building

Run in Command Prompt (`cmd.exe`) - IMPORTANT, it will not work in PowerShell or Git Bash:

```cmd
ridk enable
cd C:\work\commonmarker
where libclang.dll
set LIBCLANG_PATH=C:\Ruby40-x64\msys64\ucrt64\bin\
set RB_SYS_CARGO_PROFILE=dev
bundle exec rake compile:dev
bundle exec rake native:commonmarker gem
```

The newly built gem can be installed like this:

```console
WDAGUtilityAccount@8e95d2f3-261f-417a-8a06-5fbf6c031e74 MINGW64 /c/work/commonmarker (windows-segfault-repro)
$ gem install pkg/commonmarker-2.8.2-x64-mingw-ucrt.gem
Successfully installed commonmarker-2.8.2-x64-mingw-ucrt
1 gem installed
```

---
</details>

First of all, I found out that if I compile the extension using the `dev` profile, the segfault disappears and everything works fine. This is where reviewing https://github.com/gjtorikian/commonmarker/compare/v2.7.0...v2.8.0 helped - I found https://github.com/gjtorikian/commonmarker/commit/4b3c26663f4c5ba660d3fd0c8dc5318a5eb3b777, which enabled [LTO](https://doc.rust-lang.org/cargo/reference/profiles.html#lto) by setting `lto = true` in `[profile.release]`. (I believe that explains why the segfault did not occur in `2.7.0` and suddenly occurs in `2.8.0`).

So I enabled LTO in the `dev` profile as well (https://github.com/generalmimon/commonmarker/commit/050cb7641123f12b5927299a3a51a5f386ef3c7f):

```diff
diff --git a/Cargo.toml b/Cargo.toml
index b6457ba6..a268578f 100644
--- a/Cargo.toml
+++ b/Cargo.toml
@@ -6,6 +6,9 @@
 members = ["ext/commonmarker"]
 resolver = "2"

+[profile.dev]
+lto = true
+
 [profile.release]
 codegen-units = 1
 lto = true
```

After that, I compiled https://github.com/generalmimon/commonmarker/tree/fa8608b611c3144e507062c6dcaf795cd17afa60 with the `dev` profile, installed the `pkg/commonmarker-2.8.2-x64-mingw-ucrt.gem` file using `gem install` and ran `ruby segfault-repro.rb` in WinDbg (version 1.2603.20001.0).

WinDbg reported the segfault like this:

```
(2794.7d0): Access violation - code c0000005 (first chance)
First chance exceptions are reported before any exception handling.
This exception may be expected and handled.
commonmarker!match_at+0x422:
00007ffa`8f466b82 ffe0            jmp     rax {abababab`ababab01}
```

... and this is what the call stack looked like at that moment:

```
Frame Index	Call Site	Child-SP	Return Address
[0x0]	commonmarker!match_at+0x422 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig_sys-69.9.3/oniguruma/src/regexec.c @ 3080]	0x3e543f9080	0x7ffa8f4884ec
[0x1]	commonmarker!search_in_range+0xbac [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig_sys-69.9.3/oniguruma/src/regexec.c @ 5713]	0x3e543fa8e0	0x7ffa8f478347
[0x2]	commonmarker!onig_search_with_param+0x77 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig_sys-69.9.3/oniguruma/src/regexec.c @ 5841]	0x3e543faa10	0x7ffa8f0bbe74
[0x3]	commonmarker!onig::Regex::search_with_param<&str>+0x2f4 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig-6.5.3/src/lib.rs @ 722]	0x3e543faa80	0x7ffa8f35fb5b
[0x4]	commonmarker!syntect::parsing::regex::regex_impl::Regex::search+0x9b [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/syntect-5.3.0/src/parsing/regex.rs @ 186]	0x3e543fad00	0x7ffa8f3251c3
[0x5]	commonmarker!syntect::parsing::regex::Regex::search+0x83 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/syntect-5.3.0/src/parsing/regex.rs @ 66]	0x3e543fadd0	0x7ffa8f37113a
[0x6]	commonmarker!syntect::parsing::parser::ParseState::search+0x3aa [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/syntect-5.3.0/src/parsing/parser.rs @ 471]	0x3e543fae60	0x7ffa8f36e470
[0x7]	commonmarker!syntect::parsing::parser::ParseState::find_best_match+0xcd0 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/syntect-5.3.0/src/parsing/parser.rs @ 390]	0x3e543fb190	0x7ffa8f36ccbc
[0x8]	commonmarker!syntect::parsing::parser::ParseState::parse_next_token+0x1ac [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/syntect-5.3.0/src/parsing/parser.rs @ 277]	0x3e543fbf50	0x7ffa8f36c8ff
[0x9]	commonmarker!syntect::parsing::parser::ParseState::parse_line+0x64f [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/syntect-5.3.0/src/parsing/parser.rs @ 238]	0x3e543fc760	0x7ffa8f37bdde
[0xa]	commonmarker!syntect::html::ClassedHTMLGenerator::parse_html_for_line_which_includes_newline+0x5e [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/syntect-5.3.0/src/html.rs @ 94]	0x3e543fccf0	0x7ffa8eef55c1
[0xb]	commonmarker!comrak::plugins::syntect::SyntectAdapter::highlight_html+0x331 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/comrak-0.52.0/src/plugins/syntect.rs @ 71]	0x3e543fd060	0x7ffa8eef5115
[0xc]	commonmarker!comrak::plugins::syntect::{impl#1}::write_highlighted+0x1f5 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/comrak-0.52.0/src/plugins/syntect.rs @ 102]	0x3e543fd8c0	0x7ffa8eed1b7a
[0xd]	commonmarker!comrak::html::render_code_block<()>+0x10aa [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/comrak-0.52.0/src/html.rs @ 559]	0x3e543fda90	0x7ffa8eecc560
[0xe]	commonmarker!comrak::html::format_node_default<()>+0xa80 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/comrak-0.52.0/src/html.rs @ 345]	0x3e543fe0d0	0x7ffa8eeccfdb
[0xf]	commonmarker!comrak::html::format_document_with_formatter<()>+0x43b [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/comrak-0.52.0/src/html.rs @ 308]	0x3e543fe2e0	0x7ffa8eed59b6
[0x10]	commonmarker!comrak::html::format_document_with_plugins+0x36 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/comrak-0.52.0/src/html.rs @ 53]	0x3e543fe620	0x7ffa8ed9256d
[0x11]	commonmarker!commonmarker::commonmark_to_html+0x97d [C:/work/commonmarker/ext/commonmarker/src/lib.rs @ 151]	0x3e543fe680	0x7ffa8edc55e9
```

<details>
<summary>Hidden stack frames</summary>

```
[0x12]	commonmarker!Init_commonmarker+0x342b9	0x3e543fef40	0x7ffa8ed99d56
[0x13]	commonmarker!magnus::method::RubyFunctionCAry::call_convert_value<fn(&magnus::api::Ruby, &[magnus::value::Value]) -> core::result::Result<magnus::r_string::RString, magnus::error::Error>, core::result::Result<magnus::r_string::RString, magnus::error::Error>>+0xf6 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/magnus-0.8.2/src/method.rs @ 1398]	0x3e543fef90	0x7ffa8eda3610
[0x14]	commonmarker!Init_commonmarker+0x122e0	0x3e543ff070	0x7ffa8eda281b
[0x15]	commonmarker!core::panic::unwind_safe::{impl#25}::call_once<core::result::Result<magnus::value::Value, magnus::error::Error>, magnus::method::RubyFunctionCAry::call_handle_error::{closure_env#0}<fn(&magnus::api::Ruby, &[magnus::value::Value]) -> core::result::Result<magnus::r_string::RString, magnus::error::Error>, core::result::Result<magnus::r_string::RString, magnus::error::Error>>>+0x1b [/rustc/31fca3adb283cc9dfd56b49cdee9a96eb9c96ffd/library/core/src/panic/unwind_safe.rs @ 275]	0x3e543ff0b0	0x7ffa8eda5a5d
[0x16]	commonmarker!std::panicking::catch_unwind::do_call<core::panic::unwind_safe::AssertUnwindSafe<magnus::method::RubyFunctionCAry::call_handle_error::{closure_env#0}<fn(&magnus::api::Ruby, &[magnus::value::Value]) -> core::result::Result<magnus::r_string::RString, magnus::error::Error>, core::result::Result<magnus::r_string::RString, magnus::error::Error>>>, core::result::Result<magnus::value::Value, magnus::error::Error>>+0x3d [/rustc/31fca3adb283cc9dfd56b49cdee9a96eb9c96ffd/library/std/src/panicking.rs @ 581]	0x3e543ff100	0x7ffa8ef74fe8
[0x17]	commonmarker!Init_commonmarker+0x1e3cb8	0x3e543ff1b0	0x7ffa8ef77410
[0x18]	commonmarker!Init_commonmarker+0x1e60e0	0x3e543ff1f0	0x7ffa8ed9986c
[0x19]	commonmarker!magnus::method::RubyFunctionCAry::call_handle_error<fn(&magnus::api::Ruby, &[magnus::value::Value]) -> core::result::Result<magnus::r_string::RString, magnus::error::Error>, core::result::Result<magnus::r_string::RString, magnus::error::Error>>+0x2c [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/magnus-0.8.2/src/method.rs @ 1403]	0x3e543ff2c0	0x7ffa8ed92b77
[0x1a]	commonmarker!commonmarker::init::anon+0x17 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/magnus-0.8.2/src/method.rs @ 1586]	0x3e543ff3f0	0x7ffa9a27e5b0
[0x1b]	x64_ucrt_ruby400!rb_error_arity+0x130	0x3e543ff440	0x7ffa9a27e8ac
[0x1c]	x64_ucrt_ruby400!rb_error_arity+0x42c	0x3e543ff520	0x7ffa9a2a3db2
[0x1d]	x64_ucrt_ruby400!rb_ec_obj_respond_to+0x932	0x3e543ff580	0x7ffa9a292a30
[0x1e]	x64_ucrt_ruby400!rb_vm_exec+0x220	0x3e543ff740	0x7ffa9a05b976
[0x1f]	x64_ucrt_ruby400!rb_refinement_module_get_refined_class+0x2e56	0x3e543ff8b0	0x7ffa9a05dfd0
[0x20]	x64_ucrt_ruby400!ruby_run_node+0x70	0x3e543ffa30	0x7ff6b3fa1503
[0x21]	ruby+0x1503	0x3e543ffa90	0x7ff6b3fa10da
[0x22]	ruby+0x10da	0x3e543ffaf0	0x7ff6b3fa1456
[0x23]	ruby+0x1456	0x3e543ffb90	0x7ffada6b7374
[0x24]	KERNEL32!BaseThreadInitThunk+0x14	0x3e543ffbc0	0x7ffadae1cc91
[0x25]	ntdll!RtlUserThreadStart+0x21	0x3e543ffbf0	0x0
```
</details>

Then I spent some time isolating and minimizing the Rust code triggerring the segfault. I deleted most of the code, removed the `comrak` and `syntect` dependencies, and decided to call the `onig` methods directly, using the same arguments as if `syntect` were calling them. I ended up with this code (https://github.com/generalmimon/commonmarker/blob/44d208394e9add04f25dddf0ccef8ab0db8ff0e9/ext/commonmarker/src/lib.rs#L52-L55):

```rs
    let regex = onig::Regex::with_options("", onig::RegexOptions::REGEX_OPTION_CAPTURE_GROUP, onig::Syntax::default())
        .map_err(|e| magnus::Error::new(ruby.exception_runtime_error(), e.to_string()))?;
    let mut region = onig::Region::with_capacity(8);
    let matched = regex.search_with_param("Hello, world\n", 0, 13, onig::SearchOptions::SEARCH_OPTION_NONE, Some(&mut region), onig::MatchParam::default());
```

I also enabled the feature `print-debug` of the `onig` crate (https://github.com/generalmimon/commonmarker/blob/44d208394e9add04f25dddf0ccef8ab0db8ff0e9/ext/commonmarker/Cargo.toml#L14):

```toml
onig = { version = "6.5.3", features = ["print-debug"] }
```

That alone revealed something suspicious - the output was different with LTO enabled (`lto = true`):

```console
$ ruby segfault-repro.rb
plugins: {syntax_highlighter: {theme: "", path: ""}}
onig_search (entry point): str: 00007FFA9A15E5B0, end: 13, start: 0, range: 13
onig_search(apply anchor): end: 13, start: 0, range: 13
match_at: str: 00007FFA9A15E5B0, end: 00007FFA9A15E5BD, start: 00007FFA9A15E5B0
size: 13, start offset: 0
Segmentation fault         ruby segfault-repro.rb
```

... and with LTO disabled (i.e. after changing [`Cargo.toml:10`](https://github.com/generalmimon/commonmarker/blob/44d208394e9add04f25dddf0ccef8ab0db8ff0e9/Cargo.toml#L10) to `lto = false`, or deleting the `[profile.dev]` section):

```console
$ ruby segfault-repro.rb
plugins: {syntax_highlighter: {theme: "", path: ""}}

PATTERN: //
OPTIONS: CAPTURE_GROUP
MAX PARSE DEPTH: 2
TREE (after tune)
<empty-string:000001B4F287C600>

optimize: NONE
  anchor: []

push_mem_start: 0x0, push_mem_end: 0x0
code-length: 1
   0: end

onig_search (entry point): str: 00007FFA9A10AE80, end: 13, start: 0, range: 13
onig_search(apply anchor): end: 13, start: 0, range: 13
match_at: str: 00007FFA9A10AE80, end: 00007FFA9A10AE8D, start: 00007FFA9A10AE80
size: 13, start offset: 0
      1:       1:    0> "Hello, ..."            0: end
"true"
```

The additional output starting with `PATTERN: //` comes from initializing `onig::Regex`, so I looked in WinDbg where the `onig::Regex::with_options` call leads to:

1. LTO disabled (`lto = false`)

   ```
   [0x0]	commonmarker!onig_compile+0x6b [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig_sys-69.9.3/oniguruma/src/regcomp.c @ 7506]	0xcafc1fe3b0	0x7ffa91fafe41

     (this is right after the `fprintf(DBGFP, "\nPATTERN: /");` call)

   [0x1]	commonmarker!onig_new+0xb1 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig_sys-69.9.3/oniguruma/src/regcomp.c @ 7756]	0xcafc1fe550	0x7ffa91e43ee8
   [0x2]	commonmarker!onig::Regex::with_options_and_encoding<&str>+0x298 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig-6.5.3/src/lib.rs @ 346]	0xcafc1fe5a0	0x7ffa91e4292a
   [0x3]	commonmarker!onig::Regex::with_options+0x2a [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig-6.5.3/src/lib.rs @ 293]	0xcafc1fe720	0x7ffa91e36978
   [0x4]	commonmarker!commonmarker::commonmark_to_html+0x3e8 [C:/work/commonmarker/ext/commonmarker/src/lib.rs @ 52]	0xcafc1fe770	0x7ffa91e38919
   ```

2. LTO enabled (`lto = true`)

   ```
   [0x0]	x64_ucrt_ruby400!onig_new	0x78557feb18	0x7fffbf9b5f88
   [0x1]	commonmarker!onig::Regex::with_options_and_encoding<&str>+0x298 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig-6.5.3/src/lib.rs @ 346]	0x78557feb20	0x7fffbf9b5cda
   [0x2]	commonmarker!onig::Regex::with_options+0x2a [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig-6.5.3/src/lib.rs @ 293]	0x78557feca0	0x7fffbf8f0a48
   [0x3]	commonmarker!commonmarker::commonmark_to_html+0x3e8 [C:/work/commonmarker/ext/commonmarker/src/lib.rs @ 52]	0x78557fecf0	0x7fffbf8994e9
   ```

As you can see, here is the answer to why the extension behaves differently with and without LTO. With LTO enabled (`lto = true`), the `onig_new` symbol is resolved to `x64_ucrt_ruby400!onig_new` instead of `commonmarker!onig_new`. However, other Oniguruma symbols like `onig_search_with_param` and `search_in_range` are resolved to the `commonmarker!...` definition in both versions (we could already realize that from the above output, but here are the stack traces to back it up):

1. LTO disabled (`lto = false`)

   ```
   [0x0]	commonmarker!search_in_range+0xa2 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig_sys-69.9.3/oniguruma/src/regexec.c @ 5483]	0x99fedfe6d0	0x7ffa91d72c57

     (this is right after printing `onig_search (entry point): str: 00007FFA91D9AE80, end: 13, start: 0, range: 13`)

   [0x1]	commonmarker!onig_search_with_param+0x77 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig_sys-69.9.3/oniguruma/src/regexec.c @ 5841]	0x99fedfe820	0x7ffa91c03274
   [0x2]	commonmarker!onig::Regex::search_with_param<&str>+0x2f4 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig-6.5.3/src/lib.rs @ 722]	0x99fedfe890	0x7ffa91bf6b44
   [0x3]	commonmarker!commonmarker::commonmark_to_html+0x5b4 [C:/work/commonmarker/ext/commonmarker/src/lib.rs @ 55]	0x99fedfeb10	0x7ffa91bf8919
   ```

2. LTO enabled (`lto = true`)

   ```
   [0x0]	commonmarker!search_in_range+0xa2 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig_sys-69.9.3/oniguruma/src/regexec.c @ 5483]	0x3423bfe420	0x7ffa91a60cf7

     (this is right after printing `onig_search (entry point): str: 00007FFA91A8E5B0, end: 13, start: 0, range: 13`)

   [0x1]	commonmarker!onig_search_with_param+0x77 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig_sys-69.9.3/oniguruma/src/regexec.c @ 5841]	0x3423bfe570	0x7ffa919555b4
   [0x2]	commonmarker!onig::Regex::search_with_param<&str>+0x2f4 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig-6.5.3/src/lib.rs @ 722]	0x3423bfe5e0	0x7ffa91854774
   [0x3]	commonmarker!commonmarker::commonmark_to_html+0x5b4 [C:/work/commonmarker/ext/commonmarker/src/lib.rs @ 55]	0x3423bfe860	0x7ffa9183b039
   ```

   Note - resuming leads directly to the segfault shortly after:

   ```
   (245c.23c0): Access violation - code c0000005 (first chance)
   First chance exceptions are reported before any exception handling.
   This exception may be expected and handled.
   commonmarker!match_at+0x4a8:
   00007ffa`91a37a98 ffe0            jmp     rax {abababab`ababab01}
   ```

   ```
   [0x0]	commonmarker!match_at+0x4a8 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig_sys-69.9.3/oniguruma/src/regexec.c @ 3079]	0x3423bfb950	0x7ffa91a7408c
   [0x1]	commonmarker!search_in_range+0xc9c [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig_sys-69.9.3/oniguruma/src/regexec.c @ 5713]	0x3423bfe420	0x7ffa91a60cf7
   [0x2]	commonmarker!onig_search_with_param+0x77 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig_sys-69.9.3/oniguruma/src/regexec.c @ 5841]	0x3423bfe570	0x7ffa919555b4
   [0x3]	commonmarker!onig::Regex::search_with_param<&str>+0x2f4 [C:/Users/WDAGUtilityAccount/.cargo/registry/src/index.crates.io-1949cf8c6b5b557f/onig-6.5.3/src/lib.rs @ 722]	0x3423bfe5e0	0x7ffa91854774
   [0x4]	commonmarker!commonmarker::commonmark_to_html+0x5b4 [C:/work/commonmarker/ext/commonmarker/src/lib.rs @ 55]	0x3423bfe860	0x7ffa9183b039
   ```

---

I did not investigate much further, but I believe this paints a pretty clear picture what's going on: `x64_ucrt_ruby400!onig_new` likely initializes the first parameter `regex_t** reg` (which is an indirect pointer to a large [`struct re_pattern_buffer`](https://github.com/kkos/oniguruma/blob/94877ac5901c30cc7022f4f464da393b04cd3375/src/regint.h#L902-L945)) incorrectly in some way, resulting in `commonmarker!match_at` choking on it. (Perhaps it's not about `x64_ucrt_ruby400!onig_new`, there are two other `x64_ucrt_ruby400!*` symbols that the `onig` crate ends up calling directly: `x64_ucrt_ruby400!onig_region_resize` and `x64_ucrt_ruby400!onig_get_encoding`. However, the principle remains the same - it's clear that none of this should be happening.)

I can see two commits in https://github.com/gjtorikian/commonmarker/compare/v2.7.0...v2.8.0 (both `Co-Authored-By: Claude Opus 4.6 (1M context)`, which I find funny) that explicitly allowed this chaos of linking to wrong `x64_ucrt_ruby400!*` definitions silently without any errors or warnings during build:

- https://github.com/gjtorikian/commonmarker/commit/a5044e21f555ca042d6274d9f6c8f6ca790203a8 - Fix Windows mingw build: allow multiple Oniguruma definitions
- https://github.com/gjtorikian/commonmarker/commit/13729a4e98648a34bb30babaa8a3055eaefe42ad - Add build.rs to fix Windows mingw Oniguruma symbol collision

The linker was absolutely right to flag multiple definitions as errors. Suppressing these errors is clearly not a solution to the problem - it only makes the situation worse. Look how beautifully it tried to warn us about the `commonmarker!onig_new` vs. `x64_ucrt_ruby400!onig_new` conflict (and all the others):

```
error: linking with `gcc` failed: exit code: 1
  |
  (...)
  = note: (...)
          C:/Ruby40-x64/msys64/ucrt64/bin/../lib/gcc/x86_64-w64-mingw32/16.1.0/../../../../x86_64-w64-mingw32/bin/ld.exe:
          C:\work\commonmarker\target\debug\deps\rustcrQiTiJ\libonig_sys-7ef04e485ca413ae.rlib(2dc0da2129e760c5-regcomp.o): in function `onig_new':
          C:\Users\WDAGUtilityAccount\.cargo\registry\src\index.crates.io-1949cf8c6b5b557f\onig_sys-69.9.3/oniguruma/src/regcomp.c:7743:
            multiple definition of `onig_new';
            C:/Ruby40-x64/lib/libx64-ucrt-ruby400.dll.a(x64_ucrt_ruby400_dll_d000351.o):(.text+0x0): first defined here
          (...)
          collect2.exe: error: ld returned 1 exit status


warning: `commonmarker` (lib) generated 1 warning
error: could not compile `commonmarker` (lib) due to 1 previous error; 1 warning emitted
```

In short, I'm convinced that suppressing these errors isn't the answer, so I would strongly recommend reverting these two commits by Claude. That will bring these errors back if you try to compile with `lto = true` for Windows (which is a good thing). But with `lto = false`, these errors disappear, so that might be a good workaround when building for Windows (this effectively brings things back to the working state as in `2.7.0`). I don't know if there's a correct way to keep compiling with LTO for Windows 🤷‍♂️

--- a/ext/commonmarker/Cargo.toml
+++ b/ext/commonmarker/Cargo.toml
@@ -10,11 +10,30 @@ magnus = { version = "0.8", features = ["rb-sys"] }
 rb-sys = { version = "*", default-features = false, features = [
     "stable-api-compiled-fallback",
 ] }
-comrak = { version = "0.52", features = ["shortcodes"] }
-syntect = { version = "5.3", features = ["plist-load"] }
 typed-arena = "2.0"
 rctree = "0.6"
 rb-allocator = "0.9"
+
+syntect = { version = "5.3", default-features = false, features = [
+    "default-themes",
+    "plist-load",
+] }
+
+# Everywhere except Windows: Oniguruma, the faster engine.
+[target.'cfg(not(windows))'.dependencies]
+comrak = { version = "0.53", default-features = false, features = [
+    "shortcodes",
+    "syntect-onig",
+] }
+
+# Windows: Ruby's runtime DLL already exports Oniguruma symbols, so statically
+# linking a second copy (onig_sys) collides at link time and segfaults (xref #467).
+# syntect-fancy swaps in the pure-Rust fancy-regex backend instead.
+[target.'cfg(windows)'.dependencies]
+comrak = { version = "0.53", default-features = false, features = [
+    "shortcodes",
+    "syntect-fancy",
+] }
 
 [lib]
 name = "commonmarker"

--- a/lib/commonmarker/version.rb
+++ b/lib/commonmarker/version.rb
@@ -1,5 +1,5 @@
 # frozen_string_literal: true
 
 module Commonmarker
-  VERSION = "2.8.2"
+  VERSION = "2.8.3"
 end

